routers/api/v1/users.go

```diff
@@ -4,29 +4,56 @@ import (
 	"net/http"
 	"time"
 
-	"github.com/unicsmcr/hs_auth/services"
-
 	"github.com/unicsmcr/hs_auth/environment"
+	"github.com/unicsmcr/hs_auth/services"
 	"github.com/unicsmcr/hs_auth/utils/auth"
+	authlevels "github.com/unicsmcr/hs_auth/utils/auth/common"
 	"go.uber.org/zap"
 
 	"github.com/unicsmcr/hs_auth/routers/api/models"
 
 	"github.com/gin-gonic/gin"
 )
 
+const passwordReplacementString = "************"
+const authClaimsKeyInCtx = "auth_claims"
+
+// AuthLevelVerifierFactory creates a middleware that checks if the user's
+// auth token has the required auth level and attaches the auth claims to
+// the reques context. Will return a 401 if the auth token is invalid or has
+// an auth level that is too low
+func (r *apiV1Router) AuthLevelVerifierFactory(level authlevels.AuthLevel) func(*gin.Context) {
+	return func(ctx *gin.Context) {
+		token := ctx.GetHeader(authHeaderName)
+		claims := auth.GetJWTClaims(token, []byte(r.env.Get(environment.JWTSecret)))
+		if claims == nil || claims.TokenType != auth.Auth {
+			models.SendAPIError(ctx, http.StatusUnauthorized, "invalid token")
+			ctx.Abort()
+			return
+		} else if claims.AuthLevel < level {
+			models.SendAPIError(ctx, http.StatusUnauthorized, "you are not authorized to use this endpoint")
+			ctx.Abort()
+			return
+		}
+		ctx.Set(authClaimsKeyInCtx, claims)
+		ctx.Next()
+	}
+}
+
 // GET: /api/v1/users/
 // Response: status int
 //           error string
 //           users []entities.User
-
 func (r *apiV1Router) GetUsers(ctx *gin.Context) {
 	users, err := r.userService.GetUsers(ctx)
 	if err != nil {
 		r.logger.Error("could not fetch users", zap.Error(err))
 		models.SendAPIError(ctx, http.StatusInternalServerError, err.Error())
 		return
 	}
+	for i := 0; i < len(users); i++ {
+		users[i].Password = passwordReplacementString
```

Comment:
Is there a way to remove the password from the object instead? Or maybe don't include it in the response from `GetUsers`?

Reply:
That's a really good idea! I've modified the user model to just omit the password when it's being marshalled into JSON